content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md

```diff
@@ -0,0 +1,558 @@
+-	`query.Mascot.Animal.Name`
+
+But the top of our `vs` stack only contains one value... What do we do?
+
+We must go deeper. Cue the music from Inception, and get ready to replace `vs []reflect.Value` with `vs [][]reflect.Value`!
+
+Multiple Stacks of Values
+-------------------------
+
+That's right, to be able to deal with having potentially multiple places to unmarshal a single JSON value into, we have a slice of slices of `reflect.Value`s. Essentially, we have multiple (1 or more) `[]reflect.Value` stacks. `decoder` now looks like this:
+
+```Go
+type decoder struct {
+	tokenizer interface {
+		Token() (json.Token, error)
+	}
+
+	// Stack of what part of input JSON we're in the middle of - objects, arrays.
+	parseState []json.Delim
+
+	// Stacks of values where to unmarshal.
+	// The top of each stack is the reflect.Value where to unmarshal next JSON value.
+	//
+	// The reason there's more than one stack is because we might be unmarshaling
+	// a single JSON value into multiple GraphQL fragments or embedded structs, so
+	// we keep track of them all.
+	vs [][]reflect.Value
+}
+```
+
+We need to modify `decode` to create additional stacks whenever we encounter an embedded struct or a struct with `graphql:"... on Type"` field tag, do some additional bookkeeping to manage multiple stacks of values, check for additional error conditions if our stacks run empty. Aside from that, the same algorithm continues to work.
+
+I think getting the data structure to contain just enough information to resolve the task was the most challenging part of getting this to work. Once it's there, the rest of the algorithm details fall into place.
+
+If you'd like to learn even more of the low-level details of the implementation, I invite you to look at the [source code](https://github.com/shurcooL/graphql/blob/master/internal/jsonutil/graphql.go) of package [`github.com/shurcooL/graphql/internal/jsonutil`](https://godoc.org/github.com/shurcooL/graphql/internal/jsonutil). It should be easy to read after this post.
+
+Payoff
+------
+
+Let's quickly revisit our original GraphQL unions example that wasn't working with standard `encoding/json` unmarshaler. When we replace `json.UnmarshalJSON` with `jsonutil.UnmarshalGraphQL`, the `Name` field gets populated! That's good news, it means we didn't do all that work for nothing.
+
+See proof on [playground](https://play.golang.org/p/Xfu2mqxZ5m).
+
+`jsonutil.UnmarshalGraphQL` also takes `graphql` struct field tags into account when unmarshaling, and doesn't get mislead by `json` field tags. If there are additional GraphQL-specific tweaks to unmarshaling behavior that need to be applied in the future, they'll be easy to apply. Best part is we're reusing the rigorous JSON tokenizer of `encoding/json` and its public API, so no need to deal with maintaining a fork.
+
+Conclusion
+----------
+
+If you got this far, thanks for following along with me on this journey! I hope you enjoyed it and/or learned something new.
```

Comment:
I'd drop this sentence or put it at the very very end

Reply:
I normally avoid sentences like this, but I let my guard down because of the holiday season, and the tone of the post felt a little more casual than usual. But oh well, I'll drop it, it doesn't add much value and  dropping it makes the post shorter.

Done.